### ISSUE TITLE
automate precondition genoration

### ISSUE THREAD

**Issue by sean-stockholm** [Clojure-Intro-Course/clojure-intro-class#98]
two step process where minimal thing is humanly readable

need to incorporate `meta` to get metadata on functions

need to parse metadata to be readable by our asset_handling

**Comment by sean-stockholm** [comment]
This should be mostly correct:
`(re-defn #'clojure.core/reduce [[:f :coll] [:f nil :coll]])`
`(re-defn #'clojure.core/map [[:f :colls]])`
`(re-defn #'clojure.core/str [[:args]])`

The types so far are:
`:arg`   any argument
`:coll`   any collection
`:n`      any number
`:colls` & collections
`:str`    string
`:strs`  & strings
`:f`       function
`:fs`     & functions
`:args` any & arguments

**Comment by sean-stockholm** [comment]
That is for input:
`(pre-re-defn #'clojure.core/reduce)`
`(pre-re-defn #'clojure.core/map)`
`(pre-re-defn #'clojure.core/str)`

**Comment by elenam** [comment]
Whats' the `nil` in 3-arg reduce?

**Comment by sean-stockholm** [comment]
good point. i just noticed that too. I'm not sure what I want that to be.

**Comment by sean-stockholm** [comment]
well, I want it to be :arg.

**Comment by sean-stockholm** [comment]
Do I? what do you think? We have no reason to check arbitrary arguments, and no reason to do anything at all with a function like `#'clojure.core/str`

**Comment by sean-stockholm** [comment]
I figured out why nil was there. `:val` was missing from my hash map that went to `:arg`. although my question still remains.

**Comment by sean-stockholm** [comment]
So, it seems to work for everything that does not have pattern matching in the function definition. I'll add functionality for that later. Before that I need to add some testing. 

Currently, it will give you nil if the name of the argument is missing from the `arg-type` map. The clojure cheatsheet entries are mostly the same as the doc-strings, so you can find that missing data there. 

You can run this line of code to check it out:

`(map #(println (pre-re-defn %)) (vals (ns-publics 'clojure.core)))`

**Comment by elenam** [comment]
Looks great, thanks! I haven't tried it yet, but Tony and I will look at it on Tuesday.
